lib/Sema/TypeCheckStmt.cpp

```diff
@@ -2229,7 +2229,23 @@ static bool checkSuperInit(ConstructorDecl *fromCtor,
       fromCtor->diagnose(diag::availability_unavailable_implicit_init,
                          ctor, superclassDecl->getName());
     }
-
+    
+    // Only allowed to synthesize a throwing super.init() call if the init being
+    // checked is also throwing.
+    if (ctor->hasThrows()) {
+      // Diagnose on nonthrowing initializer
+      if (!fromCtor->hasThrows()) {
```

Comment:
Ugh, the name of this method and associated doc comment is so misleading. We should really rename it to `hasThrowsOrRethrows` or `isThrowingOrRethrowing`.